doc/serd-sort.1
.Dd October 21, 2021
.Dt SERD-SORT 1
.Os Serd
.Sh NAME
.Nm serd-sort
.Nd reorder RDF statements
.Sh SYNOPSIS
.Nm serd-sort
.Op Fl htV
.Op Fl B Ar base
.Op Fl I Ar syntax
.Op Fl O Ar syntax
.Op Fl b Ar bytes
.Op Fl c Ar collation
.Op Fl k Ar bytes
.Op Fl o Ar filename
.Op Ar input ...
.Sh DESCRIPTION
.Nm
reorders statements in RDF data by loading everything into memory then rewriting it.
By default,
a
.Dq pretty
ordering is used which is ideal for pretty-printing to Turtle or TriG.
The
.Fl c
option can be used to request a specific ordering,
which is mainly useful when emitting a line-based syntax like NTriples or NQuads in a pipeline.
.Pp
The
.Ar input
operands are processed in command-line order.
If
.Ar input
is
.Ar -
or absent,
.Nm
reads from standard input.
.Pp
The options are as follows:
.Pp
.Bl -tag -compact -width 3n
.It Fl B Ar base
Base URI, path, or
.Cm rebase
to use the output path.
See
.Xr serd-pipe 1
for details.
.Pp
.It Fl I Ar syntax
Input syntax or option:
.Cm NQuads ,
.Cm NTriples ,
.Cm TriG ,
.Cm Turtle ,
.Cm lax ,
.Cm variables ,
.Cm relative ,
or
.Cm labels .
See
.Xr serd-pipe 1
for details.
.Pp
.It Fl O Ar syntax
Output syntax or option:
.Cm empty ,
.Cm NQuads ,
.Cm NTriples ,
.Cm TriG ,
.Cm Turtle ,
.Cm ascii ,
.Cm expanded ,
.Cm verbatim ,
.Cm terse ,
or
.Cm lax .
See
.Xr serd-pipe 1
for details.
.Pp
.It Fl V
Display version information and exit.
.Pp
.It Fl b Ar bytes
I/O block size.
See
.Xr serd-pipe 1
for details.
.Pp
.It Fl c Ar collation
A specific collation (statement ordering) to use.
This can be any ordering of the characters
.Dq SPO ,
which stand for the subject, predicate, and object of statements.
Optionally,
.Dq G
can be added as the first character,
which will sort graph-first.
Concretely, the valid values are:
.Cm SPO ,
.Cm SOP ,
.Cm OPS ,
.Cm OSP ,
.Cm PSO ,
.Cm POS ,
.Cm GSPO ,
.Cm GSOP ,
.Cm GOPS ,
.Cm GOSP ,
.Cm GPSO ,
and
.Cm GPOS .
.Pp
.It Fl h
Print the command line options.
.Pp
.It Fl k Ar bytes
Parser stack size.
See
.Xr serd-pipe 1
for details.
.Pp
.It Fl o Ar filename
Write output to the given
.Ar filename
instead of stdout.
.Pp
.It Fl t
Do not write type as
.Dq a
before other properties.
Instead, rdf:type will be written in order like any other property.
.El
.Sh EXIT STATUS
.Nm
exits with a status of 0, or non-zero if an error occured.
.Sh EXAMPLES
To pretty-print a file:
.Pp
.Dl $ serd-sort -o pretty.ttl input.ttl
.Pp
To print statements ordered by predicate, subject, then object:
.Pp
.Dl $ serd-sort -c PSO input.ttl
.Sh SEE ALSO
.Bl -item -compact
.It
.Xr serd-pipe 1
.It
.Xr serd-filter 1
.It
.Lk http://drobilla.net/software/serd/
.El
.Sh STANDARDS
.Bl -item -compact
.It
.Rs
.%A W3C
.%T RDF 1.1 NQuads
.%D February 2014
.Re
.Lk https://www.w3.org/TR/n-quads/
.It
.Rs
.%A W3C
.%D February 2014
.%T RDF 1.1 NTriples
.Re
.Lk https://www.w3.org/TR/n-triples/
.It
.Rs
.%A W3C
.%T RDF 1.1 TriG
.%D February 2014
.Re
.Lk https://www.w3.org/TR/trig/
.It
.Rs
.%A W3C
.%D February 2014
.%T RDF 1.1 Turtle
.Re
.Lk https://www.w3.org/TR/turtle/
.El
.Sh AUTHORS
.Nm
is a part of serd, by
.An David Robillard
.Mt d@drobilla.net .


NEWS
serd (1.1.1) unstable; urgency=medium

  * Add SerdBuffer for mutable buffers to keep SerdChunk const-correct
  * Add SerdWorld for shared library state
  * Add extensible logging API
  * Add option for writing terse output without newlines
  * Add support for parsing variables
  * Add support for writing terse output with minimal newlines
  * Add support for xsd:float and xsd:double literals
  * Bring read/write interface closer to C standard
  * Improve error handling
  * Make nodes opaque
  * Preserve anonymous graph syntax in TriG
  * Preserve long or short quoting from input documents
  * Remove SERD_DISABLE_DEPRECATED and SERD_DEPRECATED_BY
  * Remove serd_uri_to_path()
  * Remove support for reading Turtle named inline nodes extension
  * Remove useless character counting from API
  * Rename SerdChunk to SerdStringView
  * Simplify statement flags
  * Simplify writer style options and write UTF-8 by default
  * Support writing all escapes in Turtle and TriG prefixed names
  * Use a fixed-size reader stack
  * Use char* for strings in public API

 -- David Robillard <d@drobilla.net>  Wed, 13 Jul 2022 21:55:50 +0000

serd (0.31.1) unstable; urgency=medium

  * Add Windows path separator support to serd_node_new_file_uri()
  * Add serd_reader_skip_until_byte() to public API
  * Allow SERD_API to be defined by the user
  * Avoid creating test files in the current directory
  * Avoid using ASCII grave as a quote
  * Check for POSIX features with the build system
  * Clean up and improve test suite
  * Clean up code
  * Fix crash when trying to read chunks without starting
  * Fix hang when skipping an error at EOF when lax parsing
  * Fix incorrect parsing of strange quote escape patterns
  * Fix possible hang when writing nested Turtle lists
  * Fix potential memory leaks when a write is aborted
  * Fix relative URI creation
  * Gracefully handle bad characters in Turtle blank node syntax
  * Gracefully handle bad characters in Turtle datatype syntax
  * Improve TriG pretty-printing and remove trailing newlines
  * Improve pretty-printing of lists and inline subjects
  * Improve serdi man page
  * Improve writer error handling
  * Make URI writing stricter by default
  * Override pkg-config dependency within meson
  * Remove support for writing Turtle named inline nodes extension
  * Replace duplicated dox_to_sphinx script with sphinxygen dependency
  * Test header for warnings more strictly
  * Update standard test suites

 -- David Robillard <d@drobilla.net>  Mon, 03 Apr 2023 14:48:40 +0000

serd (0.30.16) stable; urgency=medium

  * Add SERD_STATIC to pkg-config Cflags for static-only builds
  * Adopt REUSE machine-readable licensing standard
  * Allow programs to be used from subproject
  * Fix spelling mistake in serdi man page

 -- David Robillard <d@drobilla.net>  Fri, 09 Sep 2022 03:06:45 +0000

serd (0.30.14) stable; urgency=medium

  * Fix memory consumption when reading documents
  * Switch to Meson build system
  * Update README and project metadata

 -- David Robillard <d@drobilla.net>  Sun, 10 Jul 2022 16:58:43 +0000

serd (0.30.12) stable; urgency=medium

  * Fix warnings and build issues with clang 13 and VS 2019
  * Fix writing long literals with triple quotes
  * Improve documentation style
  * Support combining several BSD-style command line flags in serdi
  * Write statements with invalid URI characters in lax mode

 -- David Robillard <d@drobilla.net>  Thu, 26 May 2022 22:47:35 +0000

serd (0.30.10) stable; urgency=medium

  * Add fallback configuration if documentation theme is unavailable
  * Fix SERD_DISABLE_DEPRECATED
  * Fix building for older MacOS versions on newer MacOS
  * Fix documentation installation directory

 -- David Robillard <d@drobilla.net>  Wed, 20 Jan 2021 15:55:35 +0000

serd (0.30.8) stable; urgency=medium

  * Deprecate serd_uri_to_path()
  * Don't install API man pages
  * Fix potential memory error when serialising URIs
  * Move headers to an include directory
  * Refuse to write relative URI references to NTriples
  * Remove the need for a generated configuration header
  * Remove use of C character class functions that may use locale
  * Split up and reorganize unit tests
  * Use aligned allocation via C11 or Windows API where possible

 -- David Robillard <d@drobilla.net>  Thu, 07 Jan 2021 21:19:36 +0000

serd (0.30.6) stable; urgency=medium

  * Fix various minor warnings and other code quality issues
  * Improve reader error handling
  * Make all free functions tolerant of NULL
  * Parse prefixed names more strictly

 -- David Robillard <d@drobilla.net>  Sun, 27 Sep 2020 10:51:38 +0000

serd (0.30.4) stable; urgency=medium

  * Fix EOF handling while reading in bulk or from strings
  * Fix lax handling of string errors
  * Fix reading from a null-delimited socket

 -- David Robillard <d@drobilla.net>  Sun, 26 Apr 2020 16:04:05 +0000

serd (0.30.2) stable; urgency=medium

  * Fix GCC 4 build
  * Fix colliding blank nodes when parsing TriG
  * Fix missing parse error messages
  * Fix parsing "a" abbreviation without padding whitespace
  * Fix parsing TriG graphs with several squashed trailing dots
  * Fix resolving some URIs against base URIs with no trailing slash
  * Improve build system and CI integration
  * Improve documentation

 -- David Robillard <d@drobilla.net>  Sun, 20 Oct 2019 22:28:01 +0000

serd (0.30.0) stable; urgency=medium

  * Add NTriples test suite
  * Add basic error tolerant parsing for line-based formats
  * Add serd_free for freeing memory allocated by serd
  * Add serdi option to write ASCII output
  * Clarify errors returned by serd_env_expand()
  * Clean up testing code
  * Fix building with MSVC
  * Fix length of base64 encoded nodes
  * Fix parsing local names that end with escaped dots
  * Fix reported error when reading statements with only a blank node
  * Improve relative URI serialisation
  * Make serdi guess input syntax from extension if unspecified
  * Make serdi syntax options case-insensitive

tools/serd-sort.c
/*
  Copyright 2011-2021 David Robillard <d@drobilla.net>

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THIS SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

#include "console.h"

#include "serd/serd.h"

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

/* Application (after parsing command-line arguments) */

// All options
typedef struct {
  SerdCommonOptions  common;
  const char*        root_uri;
  const char*        input_string;
  const char*        collation;
  char* const*       inputs;
  intptr_t           n_inputs;
  SerdStatementOrder order;
  SerdDescribeFlags  flags;
} Options;

static bool
input_has_graphs(const Options opts)
{
  if (opts.common.input.syntax) {
    return serd_syntax_has_graphs(opts.common.input.syntax);
  }

  for (intptr_t i = 0u; i < opts.n_inputs; ++i) {
    if (serd_syntax_has_graphs(serd_guess_syntax(opts.inputs[i]))) {
      return true;
    }
  }

  return false;
}

// Run the tool using the given options
static SerdStatus
run(const Options opts)
{
  SerdTool app = {NULL, NULL, NULL, NULL};

  // Set up the writing environment
  SerdStatus st = SERD_SUCCESS;
  if ((st = serd_tool_setup(&app, "serd-sort", opts.common))) {
    serd_tool_cleanup(app);
    return st;
  }

  // Determine the default order to store statements in the model
  const bool               with_graphs   = input_has_graphs(opts);
  const SerdStatementOrder default_order = opts.collation ? opts.order
                                           : with_graphs  ? SERD_ORDER_GSPO
                                                          : SERD_ORDER_SPO;

  const SerdModelFlags flags =
    (SerdModelFlags)(with_graphs * SERD_STORE_GRAPHS);

  SerdModel* const model = serd_model_new(app.world, default_order, flags);

  if (!opts.collation) {
    // If we are pretty-printing, we need an O** index
    serd_model_add_index(model, SERD_ORDER_OPS);

    if (with_graphs) {
      // If we have graphs we still need the SPO index for finding subjects
      serd_model_add_index(model, SERD_ORDER_SPO);
    }
  }

  // Read all the inputs into an inserter to load the model
  SerdSink* const inserter = serd_inserter_new(model, NULL);
  if (st || (st = serd_read_inputs(app.world,
                                   opts.common,
                                   app.env,
                                   opts.n_inputs,
                                   opts.inputs,
                                   inserter))) {
    serd_tool_cleanup(app);
    return st;
  }

  // Write the model to the output
  const SerdSink* const target = serd_writer_sink(app.writer);
  if (opts.collation) {
    SerdCursor* const cursor = serd_model_begin_ordered(model, opts.order);

    serd_env_write_prefixes(app.env, target);

    for (const SerdStatement* statement = NULL;
         !st && (statement = serd_cursor_get(cursor));
         serd_cursor_advance(cursor)) {
      st = serd_sink_write_statement(target, 0u, statement);
    }

    serd_cursor_free(cursor);
  } else {
    SerdCursor* const cursor = serd_model_begin(model);

    serd_env_write_prefixes(app.env, target);

    st = serd_describe_range(cursor, target, opts.flags);

    serd_cursor_free(cursor);
  }

  if (!st) {
    st = serd_writer_finish(app.writer);
  }

  const SerdStatus cst = serd_tool_cleanup(app);
  return st ? st : cst;
}

/* Command-line interface (before setting up serd) */

static SerdStatus
parse_statement_order(const char* const string, SerdStatementOrder* const order)
{
  static const char* const strings[] = {"SPO",
                                        "SOP",
                                        "OPS",
                                        "OSP",
                                        "PSO",
                                        "POS",
                                        "GSPO",
                                        "GSOP",
                                        "GOPS",
                                        "GOSP",
                                        "GPSO",
                                        "GPOS",
                                        NULL};

  for (unsigned i = 0; strings[i]; ++i) {
    if (!strcmp(string, strings[i])) {
      *order = (SerdStatementOrder)i;
      return SERD_SUCCESS;
    }
  }

  return SERD_ERR_BAD_ARG;
}

static int
print_usage(const char* const name, const bool error)
{
  static const char* const description =
    "Reorder RDF data by loading everything into a model then writing it.\n"
    "INPUT can be a local filename, or \"-\" to read from standard input.\n\n"
    "  -B BASE_URI   Base URI or path for resolving relative references.\n"
    "  -I SYNTAX     Input syntax (turtle/ntriples/trig/nquads),\n"
    "                or option (lax/variables/relative/global/generated).\n"
    "  -O SYNTAX     Output syntax (empty/turtle/ntriples/nquads),\n"
    "                or option (ascii/expanded/verbatim/terse/lax).\n"
    "  -V            Display version information and exit.\n"
    "  -b BYTES      I/O block size.\n"
    "  -c COLLATION  An optional \"G\" then the letters \"SPO\" in any order.\n"
    "  -h            Display this help and exit.\n"
    "  -k BYTES      Parser stack size.\n"
    "  -o FILENAME   Write output to FILENAME instead of stdout.\n";

  FILE* const os = error ? stderr : stdout;
  fprintf(os, "%s", error ? "\n" : "");
  fprintf(os, "Usage: %s [OPTION]... INPUT...\n", name);
  fprintf(os, "%s", description);
  return error;
}

// Parse the option pointed to by `iter`, and advance it to the next one
static SerdStatus
parse_option(OptionIter* const iter, Options* const opts)
{
#define ARG_ERRORF(fmt, ...) \
  fprintf(stderr, "%s: " fmt, iter->argv[0], __VA_ARGS__)

  SerdStatus st = serd_parse_common_option(iter, &opts->common);
  if (st != SERD_FAILURE) {
    return st;
  }

  const char opt = iter->argv[iter->a][iter->f];
  switch (opt) {
  case 'R':
    return serd_get_argument(iter, &opts->root_uri);

  case 'V':
    return serd_print_version("serd-sort");

  case 'c':
    if (!(st = serd_get_argument(iter, &opts->collation))) {
      if ((st = parse_statement_order(opts->collation, &opts->order))) {
        ARG_ERRORF("unknown collation \"%s\"\n", opts->collation);
        return st;
      }
    }
    return st;

  case 'h':
    print_usage(iter->argv[0], false);
    return SERD_FAILURE;

  case 's':
    return serd_get_argument(iter, &opts->input_string);

  default:
    break;
  }

  ARG_ERRORF("invalid option -- '%c'\n", opt);
  return SERD_ERR_BAD_ARG;

#undef ARG_ERRORF
}

int
main(const int argc, char* const* const argv)
{
  char  default_input[]  = "-";
  char* default_inputs[] = {default_input};

  Options opts = {{"",
                   NULL,
                   4096u,
                   1048576u,
                   {SERD_SYNTAX_EMPTY, 0u, false},
                   {SERD_SYNTAX_EMPTY, 0u, false}},
                  "",
                  NULL,
                  NULL,
                  NULL,
                  0u,
                  SERD_ORDER_SPO,
                  0u};

  // Parse all command line options (which must precede inputs)
  SerdStatus st   = SERD_SUCCESS;
  OptionIter iter = {argv, argc, 1, 1};
  while (!serd_option_iter_is_end(iter)) {
    if ((st = parse_option(&iter, &opts))) {
      return (st == SERD_FAILURE) ? 0 : print_usage(argv[0], true);
    }
  }

  // Every argument past the last option is an input
  opts.inputs   = argv + iter.a;
  opts.n_inputs = argc - iter.a;
  if (opts.n_inputs + (bool)opts.input_string == 0) {
    opts.n_inputs = 1;
    opts.inputs   = default_inputs;
  }

  // Don't add prefixes to blank node labels if there is only one input
  if (opts.n_inputs + (bool)opts.input_string == 1) {
    opts.common.input.flags |= SERD_READ_GLOBAL;
  }

  return run(opts) > SERD_FAILURE;
}
